src/mc-rangeopts.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2022-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mc-rangeopts-private.h"
+
+#include "mongocrypt-private.h"
+#include "mongocrypt-util-private.h" // mc_bson_type_to_string
+
+// Common logic for testing field name, tracking duplication, and presence.
+#define IF_FIELD(Name, ErrorPrefix)                                            \
+   if (0 == strcmp (field, #Name)) {                                           \
+      if (has_##Name) {                                                        \
+         CLIENT_ERR ("%sUnexpected duplicate field '" #Name "'", ErrorPrefix); \
+         return false;                                                         \
+      }                                                                        \
+      has_##Name = true;
+
+#define END_IF_FIELD \
+   continue;         \
+   }
+
+#define CHECK_HAS(Name, ErrorPrefix)                           \
+   if (!has_##Name) {                                          \
+      CLIENT_ERR ("%sMissing field '" #Name "'", ErrorPrefix); \
+      return false;                                            \
+   }
+
+bool
+mc_RangeOpts_parse (mc_RangeOpts_t *ro,
+                    const bson_t *in,
+                    mongocrypt_status_t *status)
+{
+   bson_iter_t iter;
+   bool has_min = false, has_max = false, has_sparsity = false;
+   const char *error_prefix = "Error parsing RangeOpts: ";
+
+   BSON_ASSERT_PARAM (ro);
+   BSON_ASSERT_PARAM (in);
+   BSON_ASSERT (status || true);
+
+   *ro = (mc_RangeOpts_t){0};
+   ro->bson = bson_copy (in);
+
+   if (!bson_iter_init (&iter, ro->bson)) {
+      CLIENT_ERR ("%sInvalid BSON", error_prefix);
+      return false;
+   }
+
+   while (bson_iter_next (&iter)) {
+      const char *field = bson_iter_key (&iter);
+      BSON_ASSERT (field);
+
+      IF_FIELD (min, error_prefix)
+      ro->min = iter;
+      END_IF_FIELD
+      IF_FIELD (max, error_prefix)
+      ro->max = iter;
+      END_IF_FIELD
+      IF_FIELD (sparsity, error_prefix)
+      if (!BSON_ITER_HOLDS_INT64 (&iter)) {
+         CLIENT_ERR ("%sExpected int64 for sparsity, got: %s",
+                     error_prefix,
+                     mc_bson_type_to_string (bson_iter_type (&iter)));
+         return false;
+      };
+      ro->sparsity = bson_iter_int64 (&iter);
+      END_IF_FIELD
```

Comment:
```suggestion
      IF_FIELD (min, error_prefix)
      ro->min = iter;
      END_IF_FIELD

      IF_FIELD (max, error_prefix)
      ro->max = iter;
      END_IF_FIELD

      IF_FIELD (sparsity, error_prefix)
      if (!BSON_ITER_HOLDS_INT64 (&iter)) {
         CLIENT_ERR ("%sExpected int64 for sparsity, got: %s",
                     error_prefix,
                     mc_bson_type_to_string (bson_iter_type (&iter)));
         return false;
      };
      ro->sparsity = bson_iter_int64 (&iter);
      END_IF_FIELD
```
Whitespace to better indicate `IF_FIELD` blocks.